### PR TITLE
Distro productmap doc

### PR DESCRIPTION
**Pull request by aaannz**
## What does this PR change?

This PR adds:

- changes productmap to include 3rd level for arch, adds arch constants
- adds productmap docs
- adds distro details auto discovery
- adds api-server autodiscovery

## Test coverage
- No tests: **add explanation**
- No tests: already covered
- Unit tests were added

- [ ] **DONE**

## Links

Issue(s): https://github.com/SUSE/spacewalk/issues/23315

- [X] **DONE**

## Changelogs

Make sure the changelogs entries you are adding are compliant with https://github.com/uyuni-project/uyuni/wiki/Contributing#changelogs and https://github.com/uyuni-project/uyuni/wiki/Contributing#uyuni-projectuyuni-repository

If you don't need a changelog check, please mark this checkbox:

- [ ] No changelog needed

If you uncheck the checkbox after the PR is created, you will need to re-run `changelog_test` (see below)

# Before you merge

Check [How to branch and merge properly](https://github.com/uyuni-project/uyuni/wiki/How-to-branch-and-merge-properly)!

### CODE REVIEW

File: mgradm/cmd/distro/cp.go [hunk 1]
```diff
@@ -29,10 +28,12 @@ func umountAndRemove(mountpoint string) {
 	}
 
 	if err := utils.RunCmd("/usr/bin/sudo", umountCmd...); err != nil {
-		log.Fatal().Err(err).Msgf(L("Unable to unmount ISO image, leaving %s intact"), mountpoint)
+		log.Error().Err(err).Msgf(L("unable to unmount iso file, leaving %s intact"), mountpoint)
```
Comment (cbosdo) [inline mgradm/cmd/distro/cp.go:31]: Why are you changing the message here? It seems like a rebase issue

File: mgradm/cmd/distro/cp.go [hunk 2]
```diff
@@ -55,41 +56,16 @@ func registerDistro(connection *api.ConnectionDetails, distro *types.Distributio
...
-		return fmt.Errorf(L("distribution already exists: %s"), dstpath)
-	}
-
+func prepareSource(source string) (string, bool, error) {
 	srcdir := source
+	needremove := false
 	if strings.HasSuffix(source, ".iso") {
-		log.Debug().Msg("Source is an ISO image")
-		tmpdir, err := os.MkdirTemp("", "mgrctl")
+		log.Debug().Msg(L("Source is an iso file"))
```
Comment (cbosdo) [inline mgradm/cmd/distro/cp.go:63]: Same here the message doesn't need to be changed. We also don't localize debug messages for now.

File: mgradm/cmd/distro/cp.go [hunk 3]
```diff
@@ -98,28 +74,98 @@ func distroCp(
 			srcdir,
 		}
 		if out, err := utils.RunCmdOutput(zerolog.DebugLevel, "/usr/bin/sudo", mountCmd...); err != nil {
-			log.Debug().Msgf("Error mounting ISO image: '%s'", out)
-			return errors.New(L("unable to mount ISO image. Mount manually and try again"))
+			log.Debug().Msgf(L("Error mounting ISO image: '%s'"), out)
```
Comment (cbosdo) [inline mgradm/cmd/distro/cp.go:77]: localized debug message

File: mgradm/cmd/distro/cp.go [hunk 4]
```diff
@@ -98,28 +74,98 @@ func distroCp(
...
+			return "", needremove, fmt.Errorf(L("unable to mount iso file: %s"), out)
```
Comment (cbosdo) [inline mgradm/cmd/distro/cp.go:78]:
```suggestion
			return "", needremove, fmt.Errorf(L("Unable to mount ISO image: %s"), out)
```

File: mgradm/cmd/distro/cp.go [hunk 5]
```diff
@@ -98,28 +74,98 @@ func distroCp(
...
+	log.Info().Msgf(L("Copying distribution %s"), distro.TreeLabel)
 	if err := cnx.Copy(srcdir, "server:"+dstpath, "tomcat", "susemanager"); err != nil {
 		return fmt.Errorf(L("cannot copy %s: %s"), dstpath, err)
 	}
+	log.Info().Msgf(L("Distribution has been copied into %s"), distro.BasePath)
+	return nil
+}
 
-	log.Info().Msg(L("Distribution has been copied"))
+func getServerFqdn(flags *flagpole) (string, error) {
```
Comment (cbosdo) [inline mgradm/cmd/distro/cp.go:107]: Shouldn't this be in the API module and exposed?

File: mgradm/cmd/distro/cp.go [hunk 6]
```diff
@@ -98,28 +74,98 @@ func distroCp(
...
+
+	if err := copyDistro(srcdir, distribution, flags); err != nil {
+		return err
+	}
+
+	// Fill server FQDN if not provided, ignore error, will be hanled later
+	if flags.ConnectionDetails.Server == "" {
+		flags.ConnectionDetails.Server, _ = getServerFqdn(flags)
+		log.Debug().Msgf("Using api-server FQDN '%s'", flags.ConnectionDetails.Server)
+	}
```
Comment (cbosdo) [inline mgradm/cmd/distro/cp.go:164]: This would be useful in the `api` commands too

File: mgradm/cmd/distro/detect.go [hunk 7]
```diff
@@ -12,111 +12,144 @@ import (
...
 
 	dname := treeInfoViper.GetString("release.name")
 	dversion := treeInfoViper.GetString("release.version")
 	darch := treeInfoViper.GetString("general.arch")
-	log.Debug().Msgf("Detected distribution %s, version %s. arch %s", dname, dversion, darch)
-	details, err := getDistroFromDetails(dname, dversion, darch, channelLabel, flags)
-	if err != nil {
-		return err
-	}
+	log.Debug().Msgf(L("Detected distribution %s, version %s. arch %s"), dname, dversion, darch)
```
Comment (cbosdo) [inline mgradm/cmd/distro/detect.go:111]: localized debug message

File: mgradm/cmd/distro/distro.go [hunk 8]
```diff
@@ -5,20 +5,62 @@
...
+* <distribution version> is the version of the distribution, by default taken from '.treeinfo' file from the media. If'.treeinfo' is not found, command line option is required and used.
+* <distribution architecture> is distribution architecture, by default taken from '.treeinfo' file from the media. If '.treeinfo' is not found, command line option is required and used.
+* ChannelLabel is the channel label from Uyuni server and which is to be used for this distribution; can be overridden by command line flag.
+* InstallType is used when installer is known (for autoyast or kickstart) or use 'generic_rpm'.
+* TreeName is how the distribution will be presented in the Uyuni server UI. If not set <distribution name> is used.
+
+Build-in product map:
+
+{{ .DefaultProductMap }}
+`
```
Comment (cbosdo) [inline mgradm/cmd/distro/distro.go:61]:
```suggestion
	return L(`Auto installation distribution product mapping.

For distribution to be registered by the Uyuni server it is important to map distribution to the correct software channel.

Software channels can be named differently without any corellation to distribution name; it is then needed to allow custom distribution name to software channel mapping.

One way to set software channel is by flag --channel to the distribution copy command.

For frequent usage it is possible to write custom product mapping to the mgradm configuration file in the following format:

ProductMap:
  <distribution name>:
    <distribution version>:
      <distribution architecture>:
        ChannelLabel: <channel label>
        InstallType: <one of rhel_7|rhel_8|rhel_9|sles12generic|sles15generic|generic_rpm>
        TreeName: <custom distribution name>

Where
* <distribution name> is the name of the distribution, by default taken from '.treeinfo' file from the media. If '.treeinfo' is not found or available, command line option is required and used.
* <distribution version> is the version of the distribution, by default taken from '.treeinfo' file from the media. If'.treeinfo' is not found, command line option is required and used.
* <distribution architecture> is distribution architecture, by default taken from '.treeinfo' file from the media. If '.treeinfo' is not found, command line option is required and used.
* ChannelLabel is the channel label from Uyuni server and which is to be used for this distribution; can be overridden by command line flag.
* InstallType is used when installer is known (for autoyast or kickstart) or use 'generic_rpm'.
* TreeName is how the distribution will be presented in the Uyuni server UI. If not set <distribution name> is used.

Build-in product map:

{{ .DefaultProductMap }}
`)
```

File: mgradm/cmd/distro/distro.go [hunk 9]
```diff
@@ -31,24 +73,43 @@ func NewCommand(globalFlags *types.GlobalFlags) (*cobra.Command, error) {
...
 		Aliases: []string{"cp"},
 		RunE: func(cmd *cobra.Command, args []string) error {
 			return utils.CommandHelper(globalFlags, cmd, args, &flags, distroCp)
 		},
 	}
+	cpCmd.Flags().String("channel", "", "Set parent channel for the distribution.")
+
+	cpCmdHelp := &cobra.Command{
+		Use:   "productmap",
+		Short: "help on using custom distribution product map",
```
Comment (cbosdo) [inline mgradm/cmd/distro/distro.go:92]:
```suggestion
		Short: L("Help on using custom distribution product map"),
```

File: mgradm/cmd/distro/cp.go (hunk 1 above)
Comment (aaannz) [reply mgradm/cmd/distro/cp.go:31]: I was primarily getting rid of `Fatal` call. Reworded back, keeping error.

File: mgradm/cmd/distro/cp.go (hunk 4 above)
Comment (aaannz) [reply mgradm/cmd/distro/cp.go:78]: I was intentionally keeping error messages lowercase as I was yelled at by vscodium linter that error messages should not be capitalized.

File: mgradm/cmd/distro/cp.go (hunk 5 above)
Comment (aaannz) [reply mgradm/cmd/distro/cp.go:107]: Perhaps. Issue is that api is on both `mgrctl` and `mgradm` but only `mgradm` we expect to be run on the host server.
I know we have remote podman connection, etc. but I consider them as optional, particularly for api calls.

I am going to move it to the api module and let it fail if `--api-host` or podman connection is not available.

File: mgradm/cmd/distro/cp.go (hunk 1 above)
Comment (cbosdo) [reply mgradm/cmd/distro/cp.go:31]: well, the message is not the same :wink: No need to change the message itself.
```suggestion
		log.Error().Err(err).Msgf(L("Unable to unmount ISO image, leaving %s intact"), mountpoint)
```

File: mgradm/cmd/distro/cp.go (hunk 2 above)
Comment (cbosdo) [reply mgradm/cmd/distro/cp.go:63]:
```suggestion
		log.Debug().Msg("Source is an ISO image")
```

File: mgradm/cmd/distro/cp.go (hunk 3 above)
Comment (cbosdo) [reply mgradm/cmd/distro/cp.go:77]:
```suggestion
			log.Debug().Msgf("Error mounting ISO image: '%s'", out)
```

File: mgradm/cmd/distro/cp.go (hunk 4 above)
Comment (cbosdo) [reply mgradm/cmd/distro/cp.go:78]: My bad, the suggestion should have been:

```suggestion
			return "", needremove, fmt.Errorf(L("unable to mount ISO image: %s"), out)
```

File: mgradm/cmd/distro/cp.go (hunk 5 above)
Comment (cbosdo) [reply mgradm/cmd/distro/cp.go:107]: > Perhaps. Issue is that api is on both `mgrctl` and `mgradm` but only `mgradm` we expect to be run on the host server. I know we have remote podman connection, etc. but I consider them as optional, particularly for api calls.

Oh right I forgot.

> I am going to move it to the api module and let it fail if `--api-host` or podman connection is not available.

No need, keep it as is, I forgot about `mgrctl` not having the file access. We'll implement https://github.com/uyuni-project/uyuni-tools/issues/239 instead for it.

File: mgradm/cmd/distro/detect.go (hunk 7 above)
Comment (cbosdo) [reply mgradm/cmd/distro/detect.go:111]:
```suggestion
	log.Debug().Msgf("Detected distribution %s, version %s. arch %s", dname, dversion, darch)
```

**Pull request merged**